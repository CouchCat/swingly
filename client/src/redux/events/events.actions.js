import { EventsActionTypes } from './events.types';
import { firestore, getEnvironment } from '../../firebase/firebase.utils';
import { openSnackbar } from '../snackbar/snackbar.actions';
import { eventsToObject } from './events.utils';
import axios from 'axios';

export const fetchEvents = () => async dispatch => {
  try {
    let eventsObj = null;
    if (getEnvironment() === 'development') {
      const res = await axios.get('/api/data/mockevents');      
      eventsObj = res.data;
      console.log('Using Local Events Data!', res.data);
    } else if (getEnvironment() === 'production' || getEnvironment() === 'test') {
      const eventsSnap = await firestore.collection(`${getEnvironment()}/data/events`).orderBy('start').get();

      if (eventsSnap.docs.length) {
        eventsObj = eventsToObject(eventsSnap.docs);
      } else {
        throw new Error('No Documents found');
      }
    } else {
      throw new Error('Environement not set!');
    }
    
    // const eventsSnap = await firestore.collection(`${getEnvironment()}/data/events`).orderBy('start').get();

    // if (eventsSnap.docs.length) {
    //   eventsObj = eventsToObject(eventsSnap.docs);
    // } else {
    //   throw new Error('No Documents found');
    // }

    dispatch({
      type: EventsActionTypes.FETCH_EVENTS_SUCCESS,
      payload: eventsObj
    });
    // dispatch(openSnackbar('success', 'Events successfully fetched!'));
  } catch (error) {    
    dispatch({type: EventsActionTypes.FETCH_EVENTS_FAILED });
    dispatch(openSnackbar('error', error.message));
  }
}

export const setEventGuest = (userId, event, guestType) => async dispatch => {
  // const newEvent = {...event};
  // const eventRef = firestore.doc(`${getEnvironment()}/data/events/${event.id}`);
  // const eventSnap = await eventRef.get();
  let newGuestList = null;

  try {
    // if (eventSnap.exists) {
      // if (!event.guestList.hasOwnProperty(userId)) {
        // await eventRef.update({
        //   guestList: {
        //     ...event.guestList,
        //     userId: guestType 
        //   }
        // });

        newGuestList = {
          ...event.guests,
          [userId]: guestType
        }

        dispatch({
          type: EventsActionTypes.SET_EVENT_GUEST_SUCCESS,
          payload: { 
            eventId: event.id,
            guests: newGuestList
          }
        });
      // } else {
      //   // await eventRef.update({
      //   //   guestList: {
      //   //     ...event.guestList,
      //   //     userId: delete 
      //   //   }
      //   // });
      //   console.log('Set event guest error: ');
      // }
    // }    
  } catch (err) {
    dispatch({
      type: EventsActionTypes.SET_EVENT_GUEST_FAILED,
      payload: err
    });
  }
}

export const deleteEventGuest = (userId, event) => async dispatch => {
  // const eventRef = firestore.doc(`${getEnvironment()}/data/events/${event.id}`);
  // const eventSnap = await eventRef.get();
  let newGuestList = null;

  try {
    // if (eventSnap.exists) {
      newGuestList = {
        ...event.guests,
      };
      delete newGuestList[userId];

      dispatch({
        type: EventsActionTypes.DELETE_EVENT_GUEST_SUCCESS,
        payload: { 
          eventId: event.id,
          guests: newGuestList
        }
      });
    // }
  } catch (err) {
    dispatch({
      type: EventsActionTypes.DELETE_EVENT_GUEST_FAILED,
      payload: err
    });
  }
}
