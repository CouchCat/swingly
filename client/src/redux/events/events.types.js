export const EventsActionTypes = {
  FETCH_EVENTS_SUCCESS: 'FETCH_EVENTS_SUCCESS',
  FETCH_EVENTS_FAILED: 'FETCH_EVENTS_FAILED',
  SET_EVENT_GUEST_SUCCESS: 'SET_EVENT_GUEST_SUCCESS',
  SET_EVENT_GUEST_FAILED: 'SET_EVENT_GUEST_FAILED',
  DELETE_EVENT_GUEST_SUCCESS: 'DELETE_EVENT_GUEST_SUCCESS',
  DELETE_EVENT_GUEST_FAILED: 'DELETE_EVENT_GUEST_FAILED',
  CREATE_EVENT_SUCCESS: 'CREATE_EVENT_SUCCESS',
  CREATE_EVENT_FAILED: 'CREATE_EVENT_FAILED',
  EDIT_EVENT_SUCCESS: 'EDIT_EVENT_SUCCESS',
  EDIT_EVENT_FAILED: 'EDIT_EVENT_FAILED',
  DELETE_EVENT_SUCCESS: 'DELETE_EVENT_SUCCESS',
  DELETE_EVENT_FAILED: 'DELETE_EVENT_FAILED',
  SET_FILTER_TYPE: 'SET_FILTER_TYPE'
};