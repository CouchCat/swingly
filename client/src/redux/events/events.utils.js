import { firestore, getRootPath } from "../../firebase/firebase.utils";
import mockEvents from '../dummy-data/dummy-event-data';

export const deleteAllTestEventsInDb = async () => {
  const eventsSnap = await firestore.collection(`test/data/events`).get();

  if (eventsSnap.exists) {
    eventsSnap.forEach(doc => {
      doc.delete();
    });
  }
};

export const createMockEventsInDb = async (copies, hostObj) => {
  const dummyEventsRef = firestore.collection(`${getRootPath()}/data/events`);
  let mockEvent = mockEvents.events.id;
  mockEvent.host = hostObj;

  for (let i = 0; i < copies; i++) {
    const newDocRef = dummyEventsRef.doc();
    mockEvent.id = newDocRef.id;

    try {
      await newDocRef.set(mockEvent);
    } catch (error) {
      console.log('Add Mock Events Error!: ', error);
    }
  }
} 

export const eventsToObject = (events) => {
  return events.reduce((prev, event) => {
    prev[event.data().id] = event.data();
    return prev;
  }, {}); 
}

// export const testCall = async (text) => {
//   let rootPath = ''
//   const db = firestore;

//   if (process.env.NODE_ENV === 'production')
//     rootPath = 'production';
//   else if (process.env.NODE_ENV === 'test')
//     rootPath = 'test';
//   else if (process.env.NODE_ENV === 'development')
//     rootPath = 'development';
//   else 
//     rootPath = 'somewhere-over-the-rainbow';
  
//   const newText = text.toUpperCase() + '-' + text.toUpperCase();

//   const newDocRef = db.collection(`${rootPath}/data/text`).doc();
//   let dummData = {
//     id: newDocRef.id,
//     text: newText
//   };

//   try {
//     await newDocRef.set({
//       dummData
//     });
//   } catch (error) {
//     return error;
//   }

//   console.log('Where does this even log to?');  

//   return newText;
// }

// const tester = functions.httpsCallable('testCall');

// export const testCall = async (text) => {
//   const textTest = await tester(text);
//   console.log('tester text: ', textTest);
// }


// export const refresh = async () => {
//   const refreshStore = functions.httpsCallable('refreshStore');
//   const test = await refreshStore(true);
//   console.log('testing refresh: ! ', test);
// }

