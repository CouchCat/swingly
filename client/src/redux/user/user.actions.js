import { UserActionTypes } from "./user.types";
import {
  createUserProfileDocument,
  getCurrentUser,
  signInWithGithub,
  signInWithGoogle
} from "./user.utils";
import { auth, firestore, getEnvironment } from "../../firebase/firebase.utils";
import { openSnackbar } from "../snackbar/snackbar.actions";
import { fetchUsersArray } from "./user.utils";

const getUserWithProvider = async signInMethod => {
  switch (signInMethod) {
    case "google":
      return await signInWithGoogle();
    case "github":
      return await signInWithGithub();
    default:
      return Error("Invalid provider!");
  }
};

export const setCurrentUser = () => async dispatch => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const userAuth = await getCurrentUser();
    if (!userAuth) {
      dispatch({ type: UserActionTypes.USER_NOT_LOADING });
      return;
    }

    const userRef = await createUserProfileDocument(userAuth);
    const userSnapshot = await userRef.get();

    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_SUCCESS,
      payload: userSnapshot.data()
    });
  } catch (error) {
    dispatch({
      type: UserActionTypes.SET_CURRENT_USER_FAILED,
      payload: error
    });
  }
};

export const setUsers = () => async (dispatch) => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const usersObj = await fetchUsersArray();

    if (Object.keys(usersObj).length) {
      dispatch({
        type: UserActionTypes.SET_USERS_SUCCESS,
        payload: usersObj
      });
    } else {
      dispatch({
        type: UserActionTypes.SET_USERS_FAILED,
        payload: 'No users in DB!'
      });
    }
  } catch (error) {
    dispatch({
      type: UserActionTypes.SET_USERS_FAILED,
      payload: 'Fetch Users failed!'
    });
    dispatch(openSnackbar("error", "Fetch Users failed!"));
  }
}

export const signInWithProvider = signInMethod => async dispatch => {
  try {
    dispatch({ type: UserActionTypes.USER_LOADING });

    const { user } = await getUserWithProvider(signInMethod);
    const userRef = await createUserProfileDocument(user);
    const userSnapshot = await userRef.get();

    dispatch({
      type: UserActionTypes.SIGNIN_SUCCESS,
      payload: userSnapshot.data()
    });

    dispatch(openSnackbar("success", "You have been successfully signed in!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNIN_FAILED,
      payload: error
    });

    dispatch(openSnackbar("error", "Sign in failed!"));
  }
};

export const signOut = () => async dispatch => {
  try {
    await auth.signOut();
    dispatch({ type: UserActionTypes.SIGNOUT_SUCCESS });

    dispatch(openSnackbar("success", "You have been successfully signed out!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNOUT_FAILED,
      payload: error
    });

    dispatch(openSnackbar("error", "Sign out failed!"));
  }
};

export const signUp = values => async dispatch => {
  const usersSnap = await firestore.collection(`${getEnvironment()}/data/users`).get();
  let userExists = null;
  if (usersSnap) {
    usersSnap.forEach(doc => {
      if (doc.data().email === values.email) {
        userExists = doc.data();
      }
    });
  }

  if (!userExists) {
    try {
      const { user } = await auth.createUserWithEmailAndPassword(
        values.email,
        values.password
      );

      if (user) {
        const userAuth = await getCurrentUser();
        const userRef = await createUserProfileDocument(userAuth, {
          username: values.username
        });

        const userSnapshot = await userRef.get();
        dispatch({
          type: UserActionTypes.SIGNUP_SUCCESS,
          payload: userSnapshot.data()
        });

        await auth.signOut();

        dispatch(openSnackbar("success", "Sign up success!"));
      }
    } catch (error) {
      dispatch({
        type: UserActionTypes.SIGNUP_FAILED,
        payload: error
      });

      dispatch(openSnackbar("error", "Sign up failed!"));
    }
  } else {
    dispatch({ type: UserActionTypes.SIGNUP_FAILED });
    dispatch(openSnackbar("error", "User with given email already exists!"));
  }
};

export const signInWithEmail = values => async dispatch => {
  try {
    const { user } = await auth.signInWithEmailAndPassword(
      values.email,
      values.password
    );

    const userRef = await createUserProfileDocument(user);
    const userSnap = await userRef.get();

    dispatch({
      type: UserActionTypes.SIGNIN_SUCCESS,
      payload: userSnap.data()
    });

    dispatch(openSnackbar("success", "You have been successfully signed in!"));
  } catch (error) {
    dispatch({
      type: UserActionTypes.SIGNIN_FAILED,
      payload: error
    });

    dispatch(
      openSnackbar("error", "Incorrect email or password! Please try again.")
    );
  }
};

export const editUser = (values, currentUser) => async dispatch => {
  let newValues = {};
  if (values.username !== currentUser.username) {
    newValues.username = values.username;
  }

  if (values.description !== currentUser.description) {
    newValues.description = values.description;
  }

  if (Object.keys(newValues).length !== 0) {
    const userRef = firestore.doc(`${getEnvironment()}/data/users/${currentUser.id}`);
    const userSnap = await userRef.get();

    if (userSnap.exists) {
      try {
        const batch = firestore.batch();
        Object.entries(newValues).forEach(([key, val]) => {
          batch.update(userRef, {
            [key]: val
          });
        });
        await batch.commit();

        dispatch({
          type: UserActionTypes.EDIT_SUCCESS,
          payload: newValues
        });

        dispatch(openSnackbar("success", "Edit successful!"));
      } catch (error) {
        dispatch({
          type: UserActionTypes.EDIT_FAILED,
          payload: error
        });
        dispatch(
          openSnackbar("error", "Edit failed! User snapshot does not exist!")
        );
      }
    }
  } else {
    dispatch(openSnackbar("error", "Edit failed!"));
  }
};

export const deleteUser = () => async dispatch => {
  const user = await auth.currentUser;
  
  if (user) {
    try {
      await firestore.doc(`${getEnvironment()}/data/users/${user.uid}`).delete();
      await user.delete();
      dispatch({ type: UserActionTypes.DELETE_SUCCESS });
      dispatch(openSnackbar("success", "Delete successful!"));
    } catch (error) {
      dispatch({ 
        type: UserActionTypes.DELETE_FAILED,
        payload: error
      });
      dispatch(openSnackbar("error", "Deletion failed!"));
    }
  } else {
    dispatch({ type: UserActionTypes.DELETE_FAILED });    
    dispatch(openSnackbar("error", "Deletion failed! User not signed in"));
  }
};
