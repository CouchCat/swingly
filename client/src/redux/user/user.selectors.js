import { createSelector } from 'reselect';
import memoize from 'lodash/memoize';

const selectUser = state => state.user;

export const selectCurrentUser = createSelector(
  [selectUser],
  (user) => user.currentUser
);

export const selectUserIsLoading = createSelector(
  [selectUser],
  (user) => user.isLoading
);

export const selectUsers = createSelector(
  [selectUser],
  (user) => user.users
);

export const selectUserById = createSelector(
  [selectUsers],
  users => memoize(userId => {
    return users.hasOwnProperty(userId) ? users[userId] : false;
  })
);

