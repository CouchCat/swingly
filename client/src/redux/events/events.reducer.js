import { EventsActionTypes } from './events.types';

const INITIAL_STATE = {
  events: {},
  isLoading: false,
  filter: {
    guestFilter: ["none"],
    hostFilter: ['A8WnU1fmQnWQ3bjvI6PYro32Tjh1', 'wrlw2iC2tsOrM02iLm6OyUXfIv33']
  },
  error: null
};
const eventsReducer = (state = INITIAL_STATE, action) => {
  switch(action.type) {
    case EventsActionTypes.FETCH_EVENTS_SUCCESS:
      return {
        ...state,
        events: action.payload,
        isLoading: false,
        error: null
      }
    case EventsActionTypes.SET_EVENT_GUEST_SUCCESS:
    case EventsActionTypes.DELETE_EVENT_GUEST_SUCCESS:
      return {
        ...state,
        error: null,
        events: {
          ...state.events,
          [action.payload.eventId]: {
            ...state.events[action.payload.eventId],
            guests: {
              ...action.payload.guests
            }
          }
        }
      }
    case EventsActionTypes.CREATE_EVENT_SUCCESS:
      return {
        ...state,
        events: {
          ...state.events,
          [action.payload.eventId]: action.payload.event
        }
      }
    case EventsActionTypes.EDIT_EVENT_SUCCESS:
      return {
        ...state,
        error: null,
        events: {
          ...state.events,
          [action.payload.eventId]: {
            ...state.events[action.payload.eventId],
            ...action.payload.values
          }
        }
      }
    case EventsActionTypes.DELETE_EVENT_SUCCESS:
      const newEvents = Object.keys(state.events).reduce((prev, current) => {
        if (action.payload !== current) {
          prev[current] = state.events[current];
        }
        return prev;
      }, {});

      return {
        ...state,
        events: {
          ...newEvents
        }
      }
    case EventsActionTypes.SET_GUEST_FILTER:
      const newFilter = { ...state.filter };
      const index = newFilter.guestFilter.indexOf(action.payload)
      if (index > -1) {
        newFilter.guestFilter.splice(index, 1);
        if (newFilter.guestFilter.length === 0) {
          newFilter.guestFilter.push("none");
        }
      } else {
        newFilter.guestFilter.push(action.payload);
      }

      return {
        ...state,
        newFilter
      }
    case EventsActionTypes.SET_EVENT_GUEST_FAILED:
    case EventsActionTypes.FETCH_EVENTS_FAILED:
    case EventsActionTypes.DELETE_EVENT_GUEST_FAILED:
    case EventsActionTypes.CREATE_EVENT_FAILED:
    case EventsActionTypes.EDIT_EVENT_FAILED:
    case EventsActionTypes.DELETE_EVENT_FAILED:
      return {
        ...state,
        isLoading: false,
        error: action.payload
      }
    default: {
      return {
        ...state,
        isLoading: false,
        error: null
      }
    } 
  }
}

export default eventsReducer;
