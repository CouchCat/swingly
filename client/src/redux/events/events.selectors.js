import { createSelector } from "reselect";
import { selectCurrentUser, selectCurrentUserFollowing } from "../user/user.selectors";
import memoize from 'lodash/memoize';
import moment from 'moment';

export const sortByDate = eventsObj => {
  return Object.values(eventsObj).sort((a, b) => {
    a = a.start.toDate();
    b = b.start.toDate();

    return a > b ? 1 : a < b ? -1 : 0;
  });
}

const selectEvents = state => state.events;

export const selectAllEvents = createSelector(
  [selectEvents],
  (events) => events.events
);

export const selectEventById = createSelector(
  [selectAllEvents],
  (events) => memoize(id => {
    return events.hasOwnProperty(id) ? events[id]: null
  })
);

export const selectEventsIsLoading = createSelector(
  [selectEvents],
  (events) => events.isLoading
);

export const selectIsEventsFinishedLoading = createSelector(
  [selectAllEvents],
  events => Boolean(Object.keys(events).length)
);

export const selectFilter = createSelector(
  [selectEvents],
  (events) => events.filter
);

export const selectGuestFilter = createSelector(
  [selectFilter],
  (filter) => filter.guestFilter
);

export const selectHostFilter = createSelector(
  [selectFilter],
  (filter) => filter.hostFilter
);

export const selectSortedEvents = createSelector(
  [selectAllEvents],
  (events) => {
    return sortByDate(events);
  }
);

export const selectUpcomingEvents = createSelector(
  [selectSortedEvents],
  (events) => { 
    return events.filter(event => {
      if (moment(event.start.toDate()).isSameOrAfter(new Date(), 'day')) {
        return true
      } else {
        return false;
      }
    });
  }
);

export const selectFilteredEvents = createSelector(
  [selectSortedEvents, selectFilter, selectCurrentUser],
  (events, filter, currentUser) => {
    if (filter.guestFilter.length && currentUser) {
      return events.filter(event => {
        if (event.guests.hasOwnProperty(currentUser)) {
          if (filter.guestFilter.includes(event.guests[currentUser])) {
            return true;
          } else {
            return false;
          }
        } else {
          return false;
        }
      });
    } else {
      return events;
    }
  }
);

export const selectUpcomingEventsFilteredByGuestType = createSelector(
  [selectUpcomingEvents, selectFilter, selectCurrentUser],
  (events, filter, currentUser) => {
    if (filter.guestFilter.length) {
      return events.filter(event => {
        if (event.guests.hasOwnProperty(currentUser)) {
          if (filter.guestFilter.includes(event.guests[currentUser])) {
            return true;
          } else {
            return false;
          }
        } else {
          return false;
        }
      });
    } else {
      return events;
    }
  }
)

export const selectUpcomingFilteredEventsByHost = createSelector(
  [selectUpcomingEventsFilteredByGuestType, selectCurrentUserFollowing],
  (events, following) => {
    if (following) {
      return events.filter(event => {
        return event.hosts.some(host => {
          if (following.hasOwnProperty(host)) {
            return following[host]
          } else {
            return false;
          }
        }) 
      })
    } else {
      return events;
    }
  }
)


export const selectFilteredEventsByHost = createSelector(
  [selectFilteredEvents, selectCurrentUserFollowing],
  (eventsFiltered, following) => {
    if (following) {
      return eventsFiltered.filter(event => {
        return event.hosts.some(host => {
          if (following.hasOwnProperty(host)) {
            return following[host]
          } else {
            return false;
          }
        }) 
      })
    } else {
      return eventsFiltered;
    }
  }
)
export const selectEventsByUserId = createSelector(
  [selectUpcomingFilteredEventsByHost],
  events => memoize(userId => {
    if (userId === null) {
      return events;
    } else {
      return events.filter(event => {
        if (event.hosts.includes(userId)) {
          return true;
        } else {
          return false;
        }
      });
    }
  })
);