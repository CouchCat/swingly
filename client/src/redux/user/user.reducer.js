import { UserActionTypes } from './user.types';

const INITIAL_STATE = {
  currentUser: null,
  users: {},
  error: null,
  isLoading: false,
};

const userReducer = (state = INITIAL_STATE, action) => {
  switch(action.type) {
    case UserActionTypes.USER_LOADING:
      return {
        ...state,
        isLoading: true
      }
    case UserActionTypes.USER_NOT_LOADING:
      return {
        ...state,
        isLoading: false
      };
    case UserActionTypes.SET_CURRENT_USER_SUCCESS:
    case UserActionTypes.SIGNIN_SUCCESS:
      return {
        ...state,
        currentUser: action.payload,
        error: null,
        isLoading: false
      };
    case UserActionTypes.SET_USERS_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        users: action.payload
      }
    case UserActionTypes.SIGNUP_SUCCESS:
      return {
        ...state,
        currentUser: null,
        users: {
          ...state.users,
          [action.payload.id]: action.payload
        }
      }
    case UserActionTypes.SIGNOUT_SUCCESS: 
      return {
        ...state,
        currentUser: null
      }
    case UserActionTypes.DELETE_SUCCESS:
      const newUsers = { ...state.users };
      delete newUsers[state.currentUser];

      return {
        ...state,
        currentUser: null,
        users: newUsers,
        error: null
      };
    case UserActionTypes.EDIT_SUCCESS:
      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            ...action.payload
          }
        }
      };
    case UserActionTypes.FOLLOW_SUCCESS:
      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            following: {
              ...state.users[state.currentUser].following,
              [action.payload]: true
            }
          }
        }
      }
    case UserActionTypes.UNFOLLOW_SUCCESS:
      const newFollowing = { ...state.users[state.currentUser].following };
      delete newFollowing[action.payload];

      return {
        ...state,
        users: {
          ...state.users,
          [state.currentUser]: {
            ...state.users[state.currentUser],
            following: newFollowing
          }
        }
      }
    case UserActionTypes.SIGNUP_FAILED:
    case UserActionTypes.SET_CURRENT_USER_FAILED:
    case UserActionTypes.SET_USERS_FAILED:
    case UserActionTypes.SIGNIN_FAILED:
    case UserActionTypes.SIGNOUT_FAILED:
    case UserActionTypes.EDIT_FAILED:
    case UserActionTypes.DELETE_FAILED:
    case UserActionTypes.FOLLOW_FAILED:
    case UserActionTypes.UNFOLLOW_FAILED:
      return {
        ...state,
        error: action.payload,
        isLoading: false
      };
    default:
      return state;
  }
};

export default userReducer;