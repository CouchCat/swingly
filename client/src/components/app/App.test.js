import React from 'react';
import ConnectedApp, { App } from './App';
import { render } from "@testing-library/react";
import Root from '../../Root';
import * as mockUserActions from "../../redux/user/user.actions";
import * as mockEventsActions from "../../redux/events/events.actions";

jest.mock("../../redux/user/user.actions", () => {
  return {
    setUsers: jest.fn(() =>  () => Promise.resolve(null)),
    setCurrentUser: jest.fn(() =>  () => Promise.resolve(null)),
  }
});

jest.mock("../../redux/events/events.actions", () => {
  return {
    fetchEvents: jest.fn(() =>  () => Promise.resolve(null)),
  }
});

// jest.mock("../../redux/user/user.actions");
// jest.mock("../../redux/events/events.actions");

describe('App component', () => {
  let mockSnackbar = {
    type: null,
    text: null,
    isOpen: false
  };
  let mockIsLoadingUsers = null;
  let mockIsLoadingEvents = null;
  let mockIsUsersFinishedLoading = null;
  let mockIsEventsFinishedLoading = null;
  let mockProps = null;

  const applyMockProps = () => {
    mockProps = {
      setCurrentUser: jest.fn(),
      setUsers: jest.fn(),
      fetchEvents: jest.fn(),
      snackbar: mockSnackbar,
      isLoadingUsers: mockIsLoadingUsers,
      isLoadingEvents: mockIsLoadingEvents,
      isUsersFinishedLoading: mockIsUsersFinishedLoading,
      isEventsFinishedLoading: mockIsEventsFinishedLoading,
    }
  }

  const renderComponent = () => {
    return render(
      <Root>
        <ConnectedApp {...mockProps} />
      </Root>
    )
  };

  afterEach(() => {
    mockSnackbar = {
      type: null,
      text: null,
      isOpen: false
    };    
    mockIsLoadingUsers = null;
    mockIsLoadingEvents = null;
    mockIsUsersFinishedLoading = null;
    mockIsEventsFinishedLoading = null;

    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it("should render", () => {
    applyMockProps();
    const renderResult = renderComponent();

    expect(renderResult).toMatchSnapshot();
  });

  it('should call fetchEvents on mount only', async () => {
    applyMockProps();

    await renderComponent();
    
    expect(mockEventsActions.fetchEvents).toHaveBeenCalledTimes(1);
  });

  it('should call setCurrentUser on mount only', async () => {
    applyMockProps();

    await renderComponent();

    expect(mockUserActions.setCurrentUser).toHaveBeenCalledTimes(1);
  });

  it('should render the routes section if either user or event data is not currently loading', () => {
    mockIsLoadingUsers = false;
    mockIsLoadingEvents = false;
    mockIsUsersFinishedLoading = true;
    mockIsEventsFinishedLoading = true;
    applyMockProps();

    const renderResult = render(<Root><App {...mockProps} /></Root>);    
    
    expect(renderResult.queryByTestId("routes-section")).not.toBeNull();
  });

  it("should display the Spinner component while loading", () => {
    applyMockProps();

    const { queryByTestId } = renderComponent();

    expect(queryByTestId("spinner")).not.toBeNull();
  })
});