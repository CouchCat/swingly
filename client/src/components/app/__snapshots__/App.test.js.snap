// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render 1`] = `
"<Fragment>
  <GlobalStyleComponent />
  <Background />
  <Connect(Header) />
  <Connect(Snackbar) duration={3500} type={[undefined]} text={[undefined]} isOpen={[undefined]} />
  <Switch>
    <Route path=\\"/hi\\" exact={true} render={[Function: render]} />
    <Route path=\\"/profile\\" exact={true} component={{...}} />
    <Route path=\\"/\\" component={[Function: EventsPage]} />
  </Switch>
</Fragment>"
`;
